fix(challenge): update Statistics Calculator: Step 27 (#53196)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/65ca2d0625aa3a3201067f70.md

---
id: 65ca2d0625aa3a3201067f70
title: Step 30
challengeType: 0
dashedName: step-30
---

# --description--

> In `.forEach()`'s callback, use the `el` parameter to access the `counts` object. If the element is not in `counts`, set it to 0. Otherwise, increment the count for the element. Log out the final result of `counts`.

Reviewer: a few grammar fixes and missing backticks

```suggestion
In the callback function of your `forEach`, use the `el` parameter to access the `counts` object. If the element is not in `counts`, set it to `0`. Otherwise, increment the count for the element. 

Below your `forEach`, log out the final result of `counts`.
```

# --hints--

> Your .forEach() method should increment the count for each element. Don't forget the initialization scenario.

Reviewer: missing backticks.

```suggestion
Your `.forEach()` method should increment the count for each element. Don't forget the initialization scenario.
```

```js

> /array\.forEach\(\s*(\(\s*el\s*\)|el)\s*=>\s*\{?\s*(if\s*\(.*?\)\s*\{[^}]*\}\s*else\s*)?(counts\[\s*el\s*\]\s*=\s*(counts\[\s*el\s*\]\s*\|\|\s*0)\s*\+\s*1|counts\[\s*el\s*\]\s*=\s*counts\[\s*el\s*\]\s*\+\s*1);?\s*\}?/

Reviewer: we don't want to use a regex here because that was part of the original problem.
Here is a quote from the original issue

> I would suggest mocking the data so the camper can access it. That should also let us test the result of the forEach without using a regex. There is not only one valid way to solve this making the last regex test questionable.

Even though two possible solutions were mentioned in the issue, there are technically more. And trying to account for all of that using regex is going to be tough to maintain long term.

We should go with the suggestion left on the issue here 

> The test should work the same way, on the actual data and test the expected final object.

We should look into using different assert methods to test the final `counts` object.
That way the camper can write any valid solution and our tests will account for that 
https://www.chaijs.com/api/assert/

If you need help, you can reach out the original poster on the issue for ideas or on the contributors channels on the forum or discord. 👍

Author: Okay, I'll look into the Chai Assertion Library

Author: The parameter of `getMode()`, `numbers` is defined by the user so how to test the final `counts` object? Do I need to tell the user to enter a specific `numbers`?

Reviewer: In step 28, you have them add the console statement for getMode. You can add to the directions that they can test it out with the numbers  `4`, `4`, `2`, `5` and click on the `Calculate` button. 

But in your tests you can probably just hardcode the numbers array of 4,4,2,5 and then use that in your test.

```md
//hint text goes here

const numbersArr = [4,4,2,5]
// test for counts objects goes using the numebrsArr variable
```

Author: I am doing this
<img width="565" alt="image" src="https://github.com/freeCodeCamp/freeCodeCamp/assets/62249815/b7682f7d-a7c0-4b1f-badc-8fa5ee0733f2">
but the console says `[ReferenceError: counts is not defined]`

Author: I updated the step 30, please take a look

```

> Remember to log out the final result of `counts`

Reviewer: ```suggestion
You should have a `console.log` for your `counts` variable.
```

---
id: 65ca2d0625aa3a3201067f70
title: Step 30
challengeType: 0
dashedName: step-30
---

# --description--

> In the callback function of your `forEach`, use the `el` parameter to access the `counts` object. If the element is not in `counts`, set it to `0`. Otherwise, increment the count for the element. 
> 
> Below your `forEach`, log out the final result of `counts`.

Reviewer: The directions are missing the instructions to test it out with 4,4,2,5.  It is also missing the instructions for returning counts. Right now the start code returns array which will cause the test to fail. The instructions and tests should be update to accommodate that. 

````suggestion
In the callback function of your `forEach`, use the `el` parameter to access the `counts` object. If the element is not in `counts`, set it to `0`. Otherwise, increment the count for the element. 

Below your `forEach`, log out the final result of `counts`.

Then change your return statement to return `counts` instead of `array`.

To test this, enter the numbers `4, 4, 2, 5` and click `Calculate`. You should see the following in the console:

```plaintext
{ '2': 1, '4': 2, '5': 1 }
```
````

Author: > 65ca2d0625aa3a3201067f70.md: 18: prism-langs Code block languages should be supported by PrismJS ['plaintext' is not recognised.]

I got this error message for the line 18

# --hints--

> Your `.forEach()` method should increment the count for each element. Don't forget the initialization scenario. Enter numbers `4, 4, 2, 5` and click `Calculate` for testing.

Reviewer: ```suggestion
Your `.forEach()` method should increment the count for each element. Don't forget the initialization scenario. Enter the numbers `4, 4, 2, 5` and click `Calculate` for testing.
```

```js
const expected = {'2': 1, '4': 2, '5': 1};
assert.deepEqual(getMode([4, 4, 2, 5]), expected)
```

You should have a `console.log` for your `counts` variable.

```js

> assert.match(getMode.toString(), /console\.log\(\s*counts\s*\)/);

Reviewer: A new test should be added to make sure they changed it to return counts instead of return array.

        are from the mean, on average.
      </p>
      <p class="bold">Variance: <span id="variance"></span></p>
      <p>
        The <dfn>standard deviation</dfn> of a list of numbers is the square
        root of the variance.
      </p>
      <p class="bold">
        Standard Deviation: <span id="standardDeviation"></span>
      </p>
    </div>
  </body>
</html>
```

```css
body {
  margin: 0;
  background-color: rgb(27, 27, 50);
  text-align: center;
  color: #fff;
}

button {
  cursor: pointer;
  background-color: rgb(59, 59, 79);
  border: 3px solid white;
  color: white;
}

input {
  background-color: rgb(10, 10, 35);
  color: white;
  border: 1px solid rgb(59, 59, 79);
}

.bold {
  font-weight: bold;
}
```

```js
const getMean = (array) => array.reduce((acc, el) => acc + el, 0) / array.length;

const getMedian = (array) => {
  const sorted = array.sort((a, b) => a - b);
  const median =
    array.length % 2 === 0
      ? getMean([sorted[array.length / 2], sorted[array.length / 2 - 1]])
      : sorted[Math.floor(array.length / 2)];
  return median;
}

--fcc-editable-region--
const getMode = (array) => {
  const counts = {};

>   array.forEach(el => {})

Reviewer: ```suggestion
  array.forEach(el => {
  
  })
```

---
id: 65ca2d0625aa3a3201067f70
title: Step 30
challengeType: 0
dashedName: step-30
---

# --description--

> Inside the `array.forEach()` callback function, check if the current element is inside the `counts` object.  If the element can be found, update the value of `counts[el]` by `1`. Otherwise, assign the number `0` to `counts[el]`.

Reviewer 2: ```suggestion
Inside the `array.forEach()` callback function, check if the current element is inside the `counts` object.  If the element can be found, increment the value of `counts[el]` by `1`. Otherwise, assign the number `1` to `counts[el]`.
```

This doesn't work with `0` as the starting value when using the example code we gave in Step 26.

1. `0` is a falsy value.
2. If the number is used as a computed property name it means it was in the array at least once. So it might as well be set to `1`

---

This is the code example from the issue just for context to show the difference with code that would work with `0`

```
array.forEach((el) => {
  if (!counts[el]) {
    counts[el] = 0;
  }
  counts[el] = counts[el] + 1;
});
```

> Below your `forEach`, log out the final result of `counts`. After that, change your return statement to return `counts` instead of `array`.

Reviewer 3: ```suggestion
Change your return statement to return `counts` instead of `array`.
```

To test this, enter the numbers `4, 4, 2, 5` and click `Calculate`. You should see the following in the console:

```js
{ '2': 1, '4': 2, '5': 1 }
```

# --hints--

You should change your return statement to return `counts` instead of `array`.

```js
assert.match(getMode.toString(), /return\s+counts\s*;/);
```

You should check if the current element can be found inside the `counts` object. If it can be found, update the value of `counts[el]` by `1`. Otherwise, assign the number `1` to `counts[el]`. Enter the numbers `4, 4, 2, 5` and click `"Calculate"` for testing.

```js
const expected = {'2': 1, '4': 2, '5': 1};
assert.deepEqual(getMode([4, 4, 2, 5]), expected)
```

> You should have a `console.log` for your `counts` variable.
> 
> ```js
> assert.match(getMode.toString(), /console\.log\(\s*counts\s*\)/);
> 
> ```

Reviewer 3: ```suggestion
```

---
id: 65ca2d0625aa3a3201067f70
title: Step 30
challengeType: 0
dashedName: step-30
---

# --description--

> In the callback function of your `forEach`, use the `el` parameter to access the `counts` object. If the element is not in `counts`, set it to `0`. Otherwise, increment the count for the element. 

Reviewer: The first sentence is a little confusing and was mentioned in the original issue.

```suggestion
Inside the `array.forEach()` callback function, check if the current element is inside the `counts` object.  If the element can be found, update the value of `counts[el]` by `1`. Otherwise, assign the number `0` to `counts[el]`.
```

> Below your `forEach`, log out the final result of `counts`.
> 
> Then change your return statement to return `counts` instead of `array`.

Reviewer 4: ```suggestion
Below your `forEach`, log out the final result of `counts`. After that, change your return statement to return `counts` instead of `array`.
```

To test this, enter the numbers `4, 4, 2, 5` and click `Calculate`. You should see the following in the console:

```js
{ '2': 1, '4': 2, '5': 1 }
```

# --hints--

You should change your return statement to return `counts` instead of `array`.

```js
assert.match(getMode.toString(), /return\s+counts\s*;/);
```

> Your `.forEach()` method should increment the count for each element. Don't forget the initialization scenario. Enter the numbers `4, 4, 2, 5` and click `Calculate` for testing.

Reviewer: You are only incrementing if the element is already in the `counts` object. otherwise `counts[el]` should be set to `0`

```suggestion
You should check if the current element can be found inside the `counts` object. If it can be found, update the value of `counts[el]` by `1`. Otherwise, assign the number `0` to `counts[el]`. Enter the numbers `4, 4, 2, 5` and click `"Calculate"` for testing.
```
